python_modules/dagster/dagster/_core/execution/context/system.py

```diff
@@ -714,14 +715,72 @@ def add_output_metadata(
                     f" for dynamic output '{output_def.name}' without providing a mapping key. When"
                     " logging metadata for a dynamic output, it is necessary to provide a mapping key."
                 )
-        self._metadata_accumulator = self._metadata_accumulator.with_additional_metadata(
-            output_name=output_name, metadata=metadata, mapping_key=mapping_key
+        self._metadata_accumulator = self._metadata_accumulator.with_additional_output_metadata(
+            output_name=output_name,
+            metadata=metadata,
+            mapping_key=mapping_key,
+        )
+
+    def add_asset_metadata(
+        self,
+        metadata: Mapping[str, Any],
+        asset_key: Optional[CoercibleToAssetKey] = None,
+        partition_key: Optional[str] = None,
+    ) -> None:
+        if not self.assets_def:
+            raise DagsterInvariantViolationError(
+                "Attempted to log metadata for a non-asset computation. Only assets should be calling this function."
+            )
+        if len(self.assets_def.keys) == 0:
+            raise DagsterInvariantViolationError(
+                "Attempted to log metadata without providing asset_key, but no asset_keys"
+                " are being materialized. `context.add_asset_metadata` should only be called"
+                " when materializing assets."
+            )
+        if asset_key is None and len(self.assets_def.keys) > 1:
+            raise DagsterInvariantViolationError(
+                "Attempted to log metadata without providing asset_key, but multiple asset_keys"
+                " can potentially be materialized. Please provide an asset_key to the invocation of"
+                " `context.add_asset_metadata`."
+            )
+        asset_key = AssetKey.from_coercible(asset_key) if asset_key else self.assets_def.key
+        if asset_key not in self.assets_def.keys:
+            raise DagsterInvariantViolationError(
+                f"Attempted to log metadata for asset key '{asset_key}' that is not being materialized."
+            )
+        if partition_key:
+            if not self.assets_def.partitions_def:
+                raise DagsterInvariantViolationError(
+                    f"Attempted to log metadata for partition key '{partition_key}' without a partitions definition."
+                )
+
+            targeted_partitions = self.assets_def.partitions_def.get_partition_keys_in_range(
+                partition_key_range=self.partition_key_range
+            )
+            if partition_key not in targeted_partitions:
+                raise DagsterInvariantViolationError(
+                    f"Attempted to log metadata for partition key '{partition_key}' that is not being targeted."
```

Comment:
nit: but maybe all of these messages should say "Attempted to add metadata" instead of "Attempted to log metadata"